src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs

```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text
+{
+    internal ref partial struct ValueStringBuilder
+    {
+        internal void AppendSpanFormattable<T>(T value, string? format, IFormatProvider? provider) where T : ISpanFormattable
+        {
+            if (value.TryFormat(_chars.Slice(_pos), out int charsWritten, format, provider))
+            {
+                _pos += charsWritten;
+            }
+            else
+            {
+                Append(value.ToString(format, provider));
+            }
+        }
+    }
+}
```

Comment:
Thanks, but I didn't mean to separate it out into its own file.  I meant just move the function to the primary ValueStringBuilder.cs file.

Reply:
I've tried, but it was not so easy as some things stopped building.
Can we do this in another PR ?

Reply:
C:\Work\Github\TZ_runtime\src\libraries\Common\src\System\Text\ValueStringBuilder.cs(276,110): error CS0246: The type or namespace name 'ISpanFormattable' could not be found (are you missing a using directive or an assembly reference?) [C:\Work\Github\TZ_runtime\src\libraries\System.Text.RegularExpressions\gen\System.Text.RegularExpressions.Generator.csproj]  System.Text.Json.SourceGeneration.Roslyn3.11 -> C:\Work\Github\TZ_runtime\artifacts\bin\System.Text.Json.SourceGeneration.Roslyn3.11\Debug\netstandard2.0\System.Text.Json.SourceGeneration.dll

Reply:
Thanks.  You should be able to fix that by adding this:
```C#
    internal interface ISpanFormattable : IFormattable
    {
        bool TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format, IFormatProvider? provider);
    }
```
here:
https://github.com/dotnet/runtime/blob/931b70a51abb3261e3a19b4240f9757473d4d397/src/libraries/System.Text.RegularExpressions/gen/Stubs.cs#L35-L37
or something along those lines.

Reply:
@stephentoub 
Once I've done it - release/debug build succeeds, but I say 22 projects which are linking ValueStringBuilder.
The next problem is that once I open src\libraries\System.Private.Xml\System.Private.Xml.sln, now build of tests is failing.
Again with same exception.
Should I made the trick with stubs everywhere I find linked this cs ?

Reply:
> The next problem is that once I open src\libraries\System.Private.Xml\System.Private.Xml.sln, now build of tests is failing.
Again with same exception.

Can you share that error?

Reply:
Same exception, but for project \src\libraries\Common\tests\TestUtilities\TestUtilities.csproj
9>C:\Work\Github\TZ_runtime\src\libraries\Common\src\System\Text\ValueStringBuilder.cs(276,110,276,126): error CS0246: The type or namespace name 'ISpanFormattable' could not be found (are you missing a using directive or an assembly reference?)
9>Done building project "TestUtilities.csproj" -- FAILED.

Reply:
Ugh, that's annoying.  The common TestUtilities project is pulling in the product code PathInternal.Windows.cs which in turn is using ValueStringBuilder.cs.  We should probably fix that to remove that dependency, but that's not something to be done in this PR.  So, yes, let's add that internal stub into TestUtilities as well.  Does this show up anywhere else?  If it goes beyond those two places, we'll come up with something else.

Reply:
Something strange happens. When I build Release it is successful.
Command that I am executing is according build guide: build.cmd clr+libs -rc Release
But if I tri to build tests and to run them with : build.cmd -subset libs.tests -test -rc Release
Then I see errors, also if I open solutions in VS.
Next exception is:
1>C:\Work\Github\TZ_runtime\src\libraries\Common\src\System\Text\ValueStringBuilder.cs(276,110,276,126): error CS0246: The type or namespace name 'ISpanFormattable' could not be found (are you missing a using directive or an assembly reference?)
1>Done building project "System.Text.Encodings.Web.csproj" -- FAILED.

Reply:
You can check if this is what you have in mind.
Also other comments resolved - just the name of new file is not so good, but it is descriptive :)

Reply:
@stephentoub One more bad news :)
Using AppendSpanFormattable is slower than original implementation as we use "D4" and "D2" format. And this goes to slow formatting of int : TryFormatInt32Slow

Here are results :
SpanFormattable: 
|                          Method |      Mean |    Error |   StdDev |    Median |       Min |       Max |  Gen 0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|----------:|----------:|----------:|-------:|----------:|
|               DateTime_ToString | 129.07 ns | 1.138 ns | 1.009 ns | 129.34 ns | 125.84 ns | 129.72 ns | 0.0131 |     112 B |
|         DateTime_ToString_Local | 291.06 ns | 4.934 ns | 4.615 ns | 288.28 ns | 287.16 ns | 300.48 ns | 0.0139 |     120 B |
|   DateTime_ToString_Unspecified | 127.20 ns | 1.123 ns | 1.051 ns | 127.49 ns | 123.74 ns | 127.97 ns | 0.0134 |     112 B |
| DateTime_ToString_RoundtripKind | 123.78 ns | 0.787 ns | 0.736 ns | 123.85 ns | 121.52 ns | 124.67 ns | 0.0132 |     112 B |
|               TimeSpan_ToString |  40.70 ns | 0.105 ns | 0.098 ns |  40.67 ns |  40.53 ns |  40.88 ns | 0.0066 |      56 B |

Original implementation + only using ValueStringBuilder
|                          Method |      Mean |    Error |   StdDev |    Median |       Min |       Max |  Gen 0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|----------:|----------:|----------:|-------:|----------:|
|               DateTime_ToString |  52.40 ns | 0.230 ns | 0.204 ns |  52.44 ns |  52.07 ns |  52.83 ns | 0.0094 |      80 B |
|         DateTime_ToString_Local | 187.35 ns | 2.557 ns | 2.267 ns | 187.07 ns | 183.90 ns | 192.78 ns | 0.0099 |      88 B |
|   DateTime_ToString_Unspecified |  53.19 ns | 0.464 ns | 0.412 ns |  53.28 ns |  51.94 ns |  53.59 ns | 0.0094 |      80 B |
| DateTime_ToString_RoundtripKind |  54.61 ns | 0.370 ns | 0.328 ns |  54.67 ns |  53.57 ns |  54.93 ns | 0.0095 |      80 B |
|               TimeSpan_ToString |  55.19 ns | 0.300 ns | 0.266 ns |  55.23 ns |  54.49 ns |  55.57 ns | 0.0066 |      56 B |

You can use [Performance PR](https://github.com/dotnet/performance/pull/2243) for benchmark

Reply:
@stephentoub I've pushed changes with previous formatting for D2/4/X of integers as they are much faster. But kept changes for split of AppendSpanFormattable as they seems useful and can be used in other places. Now according benchmarks looks like we have memory + time wins. Are you okay with this changes or I should add something more ?